src/cpp/cpp_typecheck_constructor.cpp

```diff
@@ -166,7 +166,7 @@ void cpp_typecheckt::default_cpctor(
   cpp_declaratort parameter_tor;
   parameter_tor.add(ID_value).make_nil();
   parameter_tor.set(ID_name, cpp_parameter);
-  parameter_tor.type()=reference_type(nil_typet());
+  parameter_tor.type() = reference_type(typet(ID_nil));
```

Comment:
These seem worse -- keep `nil_typet` and just warn people to only use it when absolutely necessary?

Reply:
That certainly is an option, but then we need to start using it consistently (there are other places that have always been doing `typet(ID_nil)`).

Reply:
I have now added `uninitialized_typet` in the C++ front-end, which really is the only place that firmly requires that a type be set to `ID_nil`.